app/controllers/saml_idp_controller.rb

```diff
@@ -94,27 +94,13 @@ def process_invalid_authn_context
     render nothing: true, status: :bad_request
   end
 
-  def authn_context_node
-    saml_request.document.xpath(
-      '//samlp:AuthnRequest/samlp:RequestedAuthnContext/saml:AuthnContextClassRef',
-      samlp: Saml::XML::Namespaces::PROTOCOL,
-      saml: Saml::XML::Namespaces::ASSERTION)
-  end
-
   def requested_authn_context
-    return authn_context_node[0].content if authn_context_node.length == 1
-
-    logger.info 'authn_context is missing'
-    nil
-  end
-
-  def logout_response_builder
-    SamlIdp::LogoutResponseBuilder.new(
-      get_saml_response_id,
-      issuer_uri,
-      saml_response_url,
-      saml_request.request_id,
-      signature_opts)
+    if saml_request.requested_authn_context
+      saml_request.requested_authn_context
+    else
+      logger.info 'authn_context is missing'
+      nil
+    end
```

Comment:
One last comment. I have 2 suggestions here:

1. Could we use memoization so that we don't have to parse the XML document more than once? Or perhaps the memoization better belongs in the upstream gem, or both!

2. Could we remove this conditional here and have the upstream gem raise an error if the authn_context is missing, assuming it's required by the SAML spec? If it's not required, then we can remove the `else` statement here because the only thing it adds from the upstream gem is the `logger` statement, which we don't need because if it's missing, then the `auth` action will call `process_invalid_authn_context` which already logs a message when a missing or invalid authn_context is passed.